Source/DafnyCore/Backends/Rust/Dafny-compiler-rust.dfy

```diff
@@ -1736,43 +1736,80 @@ module {:extern "DCOMP"} DafnyToRustCompiler {
       readIdents := {};
       var i := 0;
       newEnv := env;
-      ghost var oldStmts := stmts;
-      var stmts := stmts; // Make it mutable
       while i < |stmts| {
         var stmt := stmts[i];
-        // Avoid maybe placebo wrapping if not necessary
-        match stmt {
-          case DeclareVar(name, optType, None) =>
-            var laterAssignmentStatus := DetectAssignmentStatus(stmts[i + 1..], name);
-            newEnv := newEnv.AddAssignmentStatus(escapeVar(name), laterAssignmentStatus);
-          case DeclareVar(name, optType, Some(InitializationValue(typ))) =>
-            var tpe := GenType(typ, GenTypeContext.default());
-            var varName := escapeVar(name);
-            var laterAssignmentStatus := DetectAssignmentStatus(stmts[i + 1..], name);
-            newEnv := newEnv.AddAssignmentStatus(varName, laterAssignmentStatus);
-          case _ =>
-
-        }
-        assume {:axiom} stmt in oldStmts;
-        var stmtExpr, recIdents, newEnv2 := GenStmt(stmt, selfIdent, newEnv, isLast && (i == |stmts| - 1), earlyReturn);
-        newEnv := newEnv2;
+        // Special case "var x: T; x = ...". We need to merge them
+        // because normally x can read a variable named x, and with two statements
+        // this is not possible.
+        if stmt.DeclareVar? && stmt.maybeValue.None? &&
+           i + 1 < |stmts| &&
+           stmts[i + 1].Assign? &&
+           stmts[i + 1].lhs.Ident? &&
+           stmts[i + 1].lhs.ident == stmt.name {
```

Comment:
Relying on this implementation detail of the Dafny IR makes me a little uneasy, but if the implementation changes then the tests you added should catch it, so I think it's a reasonable fix.

Would it make sense for the Dafny IR to represent this case more explicitly, in case we want to use shadowing like this for other backends?

Reply:
I think it would make sense for the DafnyCodeGenerator to emit Dafny IR that is suitable for shadowing immediately, that is, do not separate the declaration and the assignment upfront. But that's a bit difficult given the current way it's piggy backing on other code generators. However, that'd be the right thing to do, so that we could remove this "optimization". Thanks for your concern.